File: 163__Missing_Ranges.py
class Solution(object):
    def findMissingRanges(self, nums, lower, upper):
        """
        :type nums: List[int]
        :type lower: int
        :type upper: int
        :rtype: List[str]
        """
        start = lower
        result = []

        if len(nums) == 0:
            st = str(lower)
            if start < upper:
                st += "->" + str(upper)
            result.append(st)
            return result

        for i in range(len(nums)):
            end = nums[i] - 1
            if start <= end < upper:
                 st = str(start)
                 if start < end:
                    st += "->" + str(end)
                 result.append(st)
            start = nums[i] + 1

        if nums[-1] + 1 <= upper:
            st = str(nums[-1] + 1)
            if nums[-1] + 1 < upper:
                st += "->" + str(upper)
            result.append(st)
        return result
                    

File: Java/163__Missing_Ranges.java
class Solution {
    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        List<String> result = new ArrayList<String>();
        if (nums.length == 0) {
            if (upper == lower) result.add(String.valueOf(lower));
            else result.add(helper(lower - 1, upper + 1));
            return result;
        }
        if (nums[0] - lower > 0 || (lower < 0 && nums[0] >= 0))
            result.add(helper(lower - 1, nums[0]));
        for (int i = 0; i < nums.length - 1; i++)
            if (nums[i + 1] - nums[i] > 1 || (nums[i] < 0 && nums[i + 1] >= 0))
                result.add(helper(nums[i], nums[i + 1]));
        if (upper - nums[nums.length - 1] > 0 || (nums[nums.length - 1] < 0 && upper >= 0))
            result.add(helper(nums[nums.length - 1], upper + 1));
        return result;
    }

    private String helper(int low, int high){
        if (high - low == 2)
            return String.valueOf(high - 1);
        return String.valueOf(low + 1) + "->" + String.valueOf(high - 1);
    }
}
